### ISSUE TITLE
WARNING:tensorboardX.x2num:NaN or Inf found in input tensor.

### ISSUE THREAD

**Issue by HX-Jhao** [IDSIA/hhmarl_2D#14]
![Image](https://github.com/user-attachments/assets/f6a06bd6-8332-4395-80e0-b2978c5a6eec)
请问这个警告你在跑的时候也会出现吗？
对模型的训练有影响吗？

**Comment by ardian-selmonaj** [comment]
This warning is not very crucial, however, it can happen that training may stop at a certain time. I also encountered this warning, which is mainly due to some slight version mismatch of packages. Alternatively, you can tweak learning rate (lr), entropy coefficient (entropy_coeff) or gradient clipping (grad_clip) within `.training` in the algorithm.

**Comment by HX-Jhao** [comment]
Can you provide a version of python and other packages that are more specific?

**Comment by ardian-selmonaj** [comment]
python 3.9 or 3.10 should be used. Further, try lowering learning rate `lr=5e-5` or `lr=1e-5` and increase entropy of policy `entropy_coeff=0.1` within the training file `train_hetero.py`.

**Comment by HX-Jhao** [comment]
I need to generate a video of the battle, how do I do that?

**Comment by ardian-selmonaj** [comment]
There is not a direct function to generate a video. However,  there are two options you can try. During inference, while rendering the scene, you can either do the following:

1) keep the rendered image open (e.g. in VS Code) during the game inference, so the image will update itself and you can use any tool to do screen recording on that image. 

2) you store evey rendered image separately and generate a video out of these frames.

**Comment by duidui6666** [comment]
> 没有生成视频的直接函数。但是，您可以尝试两个选项。在推理期间，在渲染场景时，您可以执行以下作：
> 
> 1. 在游戏推理期间，保持渲染的图像处于打开状态（例如在 VS Code 中），以便图像会自行更新，并且您可以使用任何工具对该图像进行屏幕录制。
> 2. 您可以单独存储每个渲染的图像，并从这些帧中生成视频。

I tried setting the render parameter in cfg to true, but it seems that the rendered image does not display each step in a round, but displays the picture of any round in several rounds.

**Comment by ardian-selmonaj** [comment]
> > 没有生成视频的直接函数。但是，您可以尝试两个选项。在推理期间，在渲染场景时，您可以执行以下作：
> > 
> > 1. 在游戏推理期间，保持渲染的图像处于打开状态（例如在 VS Code 中），以便图像会自行更新，并且您可以使用任何工具对该图像进行屏幕录制。
> > 2. 您可以单独存储每个渲染的图像，并从这些帧中生成视频。
> 
> I tried setting the render parameter in cfg to true, but it seems that the rendered image does not display each step in a round, but displays the picture of any round in several rounds.

The rendering function def evaluate() is called after 50 training iterations and does only two evaluation episodes. You can change the range of the for loop in def make_checkpoint() to have more evaluation episodes.
